### ISSUE TITLE
Clarify where global constraints are coming from

### ISSUE THREAD

**Issue by bergmark** [haskell/cabal#2643]
Example:
```
rejecting: base-4.7.0.1/installed-c64..., 4.8.0.0, 4.7.0.2, 4.7.0.1, 4.7.0.0,
4.6.0.1, 4.6.0.0, 4.5.1.0, 4.5.0.0, 4.4.1.0, 4.4.0.0, 4.3.1.0, 4.3.0.0,
4.2.0.2, 4.2.0.1, 4.2.0.0, 4.1.0.0, 4.0.0.0 (global constraint requires <4)
```

There are several ways this global constraint could have been added
* User level ~/.cabal/config
* Sandbox config
* `--constraint`

It would be very helpful if the error message would mention where the constraint came from.

**Comment by gregwebs** [comment]
Clarifying the definition of `global constraint` would help. The meaning is that it did not come from a `.cabal` file? I would call it a `configuration constraint` (not from a `.cabal` file) vs. a `package constraint` (from a `.cabal` file).

The current name `global` does not make sense to me if it came from a `.cabal.config` or `cabal.sandbox.config` file that is local to a sandbox. Better that `configuration constraint` would be able to say `sandbox constraint` and best would be `constraint from cabal.sandbox.config`.

**Comment by kosmikus** [comment]
Yes, I'm in support of this. To the solver, a constraint is "global" essentially if it's a constraint the solver knows nothing about, i.e., it's an input to the solver. It's in principle easy to tag all input constraints with additional information about their actual origin (command line, config file, sandbox, whatever).

This sounds a relatively easy improvement for someone to work on who wants to start contributing to cabal-install.

**Comment by grayjay** [comment]
I started looking into this issue, and I have some questions:

* Several configure flags create constraints, including "constraints", "flags", "tests", and "benchmarks".  These flags can each be specified on the command line or in any config file.  Is it important to track the source of all constraints, or just the ones introduced with the "constraints" flag?  Tracking flags other than "constraints" seems like a significantly larger change.
* Does it make sense to store extra information, like the source of a flag, directly on an attribute of `D.C.Setup.ConfigExFlags`?

**Comment by kosmikus** [comment]
I think general "constraint"s are most important. Obviously, the more precise information we have, the better the error messages are, but any relative improvement is already good.

Just in terms of use cases causing the most confusion in practice, I think we have:

   * targets and preferences added implicitly when building in a sandbox, in order to try to ensure that sandboxes are always consistent

   * preferences added via the Hackage preferences file

   * constraints built into cabal-install

   * constraints and packages determined by a cabal config file

Yes, I think it makes sense to add that info to `ConfigExFlags`, as I think they're used later in `planPackages` to actually extract the constraints and preferences.

**Comment by grayjay** [comment]
@kosmikus Thanks for the suggestions.  How should we show the source of preferences?  As I understand it, preferences never prevent cabal from finding an install plan, so they aren't associated with an error message.  Would it be useful to add the source of preferences to the debug message that lists all preferences, or would users want to see how the preferences affect the result?

**Comment by kosmikus** [comment]
@grayjay Sorry, yes. I wasn't thinking clearly. Preferences don't typically lead to errors. I do have ideas how one can provide feedback about preferences, but I suggest to keep that separate from this ticket for now. If we've solved this and you'd like to contribute more after that, I'm happy to try to tackle the preferences next. So I'd focus on the constraints first, and sorry for the confusion.

**Comment by grayjay** [comment]
@kosmikus No problem.  I'll just track the constraints for now.

**Comment by 23Skidoo** [comment]
Fixed by #2727.
